### ISSUE TITLE
`dagger install`  polish

### ISSUE THREAD

**Issue by sipsma** [dagger/dagger#6605]
Follow ups from https://github.com/dagger/dagger/pull/6533 around install (and managing dependencies generally):

- [ ] add support for `dagger update` to update existing deps
   - Currently this implicitly happens if you `dagger install` an existing dep at a different version, but it would be better to have an explicit command for it.
   - We could then consider whether we want `dagger install` to fail if dep at different version already exists or to continue supporting the implicit upgrade
- [ ] add support for `dagger uninstall`
   - Currently users have to hand-edit `dagger.json` to do this

Additional things worth considering too:
- `dagger install --dry-run` (if worth effort in short term, otherwise can backlog)
- consider a more explicit syntax for install, re: https://github.com/dagger/dagger/pull/6533#issuecomment-1922358536
   - I also think the convention suggested of doing `dagger -m MOD install dep` may suffice instead

---

Tentatively adding to `v0.10` milestone since these are pretty straightforward and would be very nice to have by then

**Comment by hanshal101** [comment]
Hey @sipsma , this will be my first issue that I would work on.
Can I pick this issue?

**Comment by helderco** [comment]
I think the design is pretty straightforward, we appreciate the help. 🎉

**Comment by rajatjindal** [comment]
Hi @jedevc @sipsma @marcosnils 

I am picking up the task "add support for dagger uninstall. Currently users have to hand-edit dagger.json to do this." out of this. Kindly let me know if that is ok.

thanks
Rajat Jindal

**Comment by helderco** [comment]
> I am picking up the task "add support for dagger uninstall. Currently users have to hand-edit dagger.json to do this." out of this. Kindly let me know if that is ok.

Yes, thanks!

**Comment by jedevc** [comment]
In terms of `dagger upgrade` we can use the functionality added in https://github.com/dagger/dagger/pull/8587 - when you run `dagger upgrade` we should erase-and-recalculate the pins.

**Comment by jedevc** [comment]
> Do we want to support updating just one dependency or everything in the dependencies list?

`dagger upgrade` by itself should upgrade all dependencies - but I guess it would be nice to upgrade a single named dependency (similar to how `uninstall` selects a dep).

I'm thinking that `dagger upgrade` should just re-fetch the `source` and potentially update the `pin` if it's changed since then - but it shouldn't try and modify the `source`.

So:
- If a user installs `@main`, then we re-fetch the `main` branch
- If a user installs `@v1.2.3`, then we re-fetch, but the tag probably hasn't changed (assuming well-behaved semver)
- If a user installs `@v1.2`, then we re-fetch, and maybe the tag/branch has been updated to `v1.2.4`.
	- This is tricky. Users shouldn't have to manage release branches themselves - we should auto-magically handle this like so: https://github.com/dagger/dagger/blob/e9d95a257aa2a1db779a547fee9ece06a884b445/core/schema/git.go#L456
	- But, then instead of storing the tag we got in the `source` field, we'd probably want to store the "selector"?
	- I would leave this as an extension task for later.

@jpadams @kpenfound what do you think of the above proposal?

**Comment by rajatjindal** [comment]
thanks you @jedevc. In addition to above, I was thinking of following:

- allow updating single dependency. default should be all. (covered by you above)
- only allow if the dependency is already installed. error out otherwise.
- only allow updates to newer than current version? flag to allow downgrade? OR keep it simple and just install what is requested.
- if no version specified, use latest (main? or latest tag?)
- I am thinking this is valid only when ModuleSource is Git. should probably error out for local dependency.

**Comment by jedevc** [comment]
> * if no version specified, use latest (main? or latest tag?)

I don't think you should be able to upgrade to a specific version. That functionality is already covered by the `dagger install` command. To upgrade to a specific version, I think you should `dagger install` it with the specific url.

Maybe the `dagger upgrade` command could *also* do this? But not convinced, duplicate functionality in the DX is generally a smell.

> * I am thinking this is valid only when ModuleSource is Git. should probably error out for local dependency.

Sounds good, that makes sense. If you try and `dagger upgrade` with no args, it should just skip over local module sources.

**Comment by rajatjindal** [comment]
> I don't think you should be able to upgrade to a specific version. That functionality is already covered by the dagger install command. To upgrade to a specific version, I think you should dagger install it with the specific url.

hmm, I was actually thinking to error out if user tries to install a dependency which is already a dependency. but now that you  mentioned it, I feel "dagger install" to be more natural when installing a specific version. 

having said that, I think we should let user update to a specific version using "dagger update" as well. (unless implementations turns out to be different code paths in engine, i think it should be ok to support this in both commands).
